chalice/deploy/packager.py

```diff
@@ -190,21 +217,416 @@ def inject_latest_app(self, deployment_package_filename, project_dir):
+            # From the wheel spec (PEP 427)
+            # {distribution}-{version}(-{build tag})?-{python tag}-{abi tag}-
+            # {platform tag}.whl
+            name, version = self.filename.split('-')[:2]
+        else:
+            info_fetcher = SDistMetadataFetcher(osutils=self._osutils)
+            sdist_path = self._osutils.joinpath(self._directory, self.filename)
+            name, version = info_fetcher.get_package_name_and_version(
+                sdist_path)
+        return name, version
+
+
+class SDistMetadataFetcher(object):
+    """This is the "correct" way to get name and version from an sdist."""
+    # https://git.io/vQkwV
+    _SETUPTOOLS_SHIM = (
+        "import setuptools, tokenize;__file__=%r;"
+        "f=getattr(tokenize, 'open', open)(__file__);"
+        "code=f.read().replace('\\r\\n', '\\n');"
+        "f.close();"
+        "exec(compile(code, __file__, 'exec'))"
+    )
+
+    def __init__(self, osutils=None):
+        # type: (OSUtils) -> None
+        if osutils is None:
+            osutils = OSUtils()
+        self._osutils = osutils
+
+    def _parse_pkg_info_file(self, filepath):
+        # type: (str) -> Message
+        # The PKG-INFO generated by the egg-info command is in an email feed
+        # format, so we use an email feedparser here to extract the metadata
+        # from the PKG-INFO file.
+        data = self._osutils.get_file_contents(filepath, binary=False)
+        parser = FeedParser()
+        parser.feed(data)
+        return parser.close()
+
+    def _generate_egg_info(self, package_dir):
+        # type: (str) -> str
+        setup_py = self._osutils.joinpath(package_dir, 'setup.py')
+        script = self._SETUPTOOLS_SHIM % setup_py
+
+        cmd = [sys.executable, '-c', script, '--no-user-cfg', 'egg_info']
+        egg_info_dir = self._osutils.joinpath(package_dir, 'egg-info')
+        self._osutils.makedirs(egg_info_dir)
+        cmd += ['--egg-base', 'egg-info']
+        p = subprocess.Popen(cmd, cwd=package_dir,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        p.communicate()
+        self._osutils.joinpath(egg_info_dir)
+        info_contents = self._osutils.get_directory_contents(egg_info_dir)
+        assert len(info_contents) == 1
+        pkg_info_path = self._osutils.joinpath(
+            egg_info_dir, info_contents[0], 'PKG-INFO')
+        return pkg_info_path
+
+    def _unpack_sdist_into_dir(self, sdist_path, unpack_dir):
+        # type: (str, str) -> str
+        if sdist_path.endswith('.zip'):
+            self._osutils.unpack_zipfile(sdist_path, unpack_dir)
+        elif sdist_path.endswith('.tar.gz'):
+            self._osutils.unpack_tarfile(sdist_path, unpack_dir)
+        else:
+            raise InvalidSourceDistributionNameError(sdist_path)
+        # There should only be one directory unpacked.
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        assert len(contents) == 1
```

Comment:
There should be no case where there is any customer impact from this. The line below the assert operates under the assumption that the upstream code is generating a directory with exactly one entry. This assertion acts as a dev-time safeguard since the tests may pass by happenstance depending on which file in the directory is opened with `[0]`, and it is not a useful thing to test since it is an implementation detail. 

If anyone violates that assumption during development of a feature or refactoring the assertion will explode and point them to what and where that assumption was. Then if it's no longer valid they can remove it and modify the code below it to work with their upstream code. Or they can fix their code to not break the assumption.